docs/website/docs/developers/design-docs/hip-hal-driver.md

```diff
@@ -0,0 +1,263 @@
+not yet implemented in the HIP HAL driver.
+
+### Semaphore
+
+The IREE HAL uses semaphores to synchronize work between host CPU threads and
+device GPU streams. It's a unified primitive that covers all directions--host
+to host, host to device, device to host, and device to device, and allows
+flexible signal and wait ordering--signal before wait, or wait before signal.
+There is no limit on the number of waits of the same value too.
+
+The core state of a HAL semaphore consists of a monotonically increasing 64-bit
+integer value, which forms a timeline--signaling the semaphore to a larger
+value advances the timeline and unblocks work waiting on some earlier values.
+The semantics closely mirrors
+[Vulkan timeline semaphore][vulkan-timeline-semaphore].
+
+In HIP, there is no direct equivalent primitives providing all the capabilities
+needed by the HAL semaphore abstraction. Therefore, to implement the support,
+we need to leverage multiple native CPU or HIP primitives under the hood.
+
+#### `hipEvent_t` capabilities
+
+The main synchronization mechanism is HIP event--`hipEvent_t`.
+As a functionality and integration baseline, we use `hipEvent_t` to implement the
+IREE HAL semaphore abstraction.
+
+`hipEvent_t` natively supports the following capabilities:
+
+* State: binary; either unsignaled or signaled. There can exist multiple
+  waits (e.g., via `hipEventSynchronize()` or `hipGraphAddEventWaitNode()`) for
+  the same `hipEvent_t` signal (e.g., via `hipEventRecord()` or
+  `hipGraphAddEventRecordNode()`).
+* Ordering: must be signal before wait. Waiting before signal would mean
+  waiting an empty set of work, or previously recorded work.
+* Direction: device to device, device to host.
+
+We need to fill the remaining capability gaps. Before going into details,
+the overall approach would be to:
+
+* State: we need a 64-bit integer value timeline. Given the binary state of
+  a `hipEvent_t`, each `hipEvent_t` would just be a "timepoint" on the timeline.
+* Ordering: we need to defer releasing the workload to the GPU until the
+  semaphore waits are reached on the host, or we can have some device
+  `hipEvent_t` to wait on.
+* Direction: host to host and host to device is missing; we can support that
+  with host synchronization mechanisms.
+
+#### Signal to wait analysis
+
+Concretely, for a given HAL semaphore, looking at the four directions:
+
+##### CPU signal
+
+A CPU thread signals the semaphore timeline to a new value.
+
+If there are CPU waits, it is purely on the CPU side. We just need to use common
+CPU notification mechanisms. In IREE we have `iree_event_t` wrapping various
+low-level OS primitives for it. So we can just use that to represent a wait
+timepoint. We need to keep track of all CPU wait timepoints in the timeline.
+After a new signaled value, go through the timeline and notify all those waiting
+on earlier values.
+
+If there are GPU waits, given that there are no way we can signal a `hipEvent_t` on
```

Comment:
lint check wants this wrapped at 80 characters: https://github.com/openxla/iree/actions/runs/8479099465/job/23232540154?pr=16931#step:4:5